iib/web/models.py

```diff
@@ -190,6 +192,93 @@ def add_state(self, state, state_reason):
 
+    def to_json(self, verbose=True):
+        def _state_to_json(state):
+            return {
+                'state': RequestStateMapping(state.state).name,
+                'state_reason': state.state_reason,
+                'updated': state.updated.isoformat() + 'Z',
+            }
+
+        states = [_state_to_json(state) for state in self.states]
+        # Reverse the list since the latest states should be first
+        states = list(reversed(states))
+        latest_state = states[0]
+
+        rv = {
+            'id': self.id,
+            'arches': [arch.name for arch in self.index_image.architectures],
+            'binary_image': self.binary_image.pull_specification,
+            'binary_image_resolved': self.binary_image_resolved.pull_specification,
+            'bundles': [bundle.pull_specification for bundle in self.bundles],
+            'from_index': self.from_index.pull_specification,
+            'from_index_resolved': self.from_index_resolved,
+            'index_image': self.index_image.pull_specification,
+            'state': latest_state,
+            'user': self.user.username,
+        }
+
+        if verbose:
+            rv['state_history'] = states
+
+        return rv
+
+    @classmethod
+    def from_json(cls, kwargs):
+        # Validate all required parameters are present
+        required_params = {'bundles', 'binary_image'}
+        optional_params = {'from_index', 'add_arches', 'cnr_token'}
+        missing_params = required_params - set(kwargs.keys()) - optional_params
+        if missing_params:
+            raise ValidationError(
+                'Missing required parameter(s): {}'.format(', '.join(missing_params))
+            )
+
+        # Don't allow the user to set arbitrary columns or relationships
+        invalid_params = set(kwargs.keys()) - required_params - optional_params
+        if invalid_params:
+            raise ValidationError(
+                'The following parameters are invalid: {}'.format(', '.join(invalid_params))
+            )
+
+        request_kwargs = deepcopy(kwargs)
+
+        # Validate bundles are correctly provided
+        bundles = request_kwargs.pop('bundles')
+        if not isinstance(bundles, list) or len(bundles) == 0:
+            raise ValidationError('"bundles" should be a non-empty array')
+        request_kwargs['bundles'] = [Image(pull_specification=bundle) for bundle in bundles]
+
+        # Validate binary_image is correctly provided
+        binary_image = request_kwargs.pop('binary_image')
+        if not isinstance(binary_image, str):
+            raise ValidationError('"binary_image" should be a non-empty string')
+        request_kwargs['binary_image'] = Image(pull_specification=binary_image)
+
+        # Validate from_index and cnr_token if specified
+        for param in ('from_index', 'cnr_token'):
+            param_value = request_kwargs.pop(param)
+            if param_value and not isinstance(param_value, str):
```

Comment:
This should probably verify it's not an empty string.

Reply:
I'm sorry I didn't get this change request

Reply:
With your current logic, an exception will not be raised if `from_index` is an empty string.

```python
>>> param_value = ''
>>> if param_value and not isinstance(param_value, str):
...     print('invalid')
... else:
...     print('not invalid')
... 
not invalid
```

I think you could simplify this code by adding
```python
if param_value is None:
    continue
```

Reply:
I don't think an exception will be raised when `from_index` is an empty string as shown by your example itself. An exception will only be raised when param_value is a non-empty value and not a string. Am I missing something?

Reply:
sorry I misread your comment. we want the exception to be raised? I don't see why we should raise an exception when an empty string is specified.

Reply:
We decided that an empty string should be treated as null.